tests/lib/util/node-event-generator.js

```diff
@@ -11,41 +11,273 @@
+            ]
+        );
+
+        assertEmissions(
+            "[foo, 5, foo]",
+            ["Identifier + Literal"],
+            ast => [["Identifier + Literal", ast.body[0].expression.elements[1]]]  // 5
+        );
+
+        assertEmissions(
+            "[foo, {}, 5]",
+            ["Identifier + Literal", "Identifier ~ Literal"],
+            ast => [["Identifier ~ Literal", ast.body[0].expression.elements[2]]]  // 5
+        );
+
+        assertEmissions(
+            "foo;",
+            [
+                "*",
+                ":not(*)",
+                "Identifier",
+                "ExpressionStatement > *",
+                "ExpressionStatement > Identifier",
+                "ExpressionStatement > [name='foo']",
+                "Identifier, ReturnStatement",
+                "FooStatement",
+                "[name = 'foo']",
+                "[name='foo']",
+                "Identifier[name='foo']",
+                "[name='foo'][name.length=3]",
+                ":not(Program, ExpressionStatement)",
+                ":not(Program, Identifier) > [name.length=3]"
+            ],
+            ast => [
+                ["*", ast], // Program
+                ["*", ast.body[0]], // ExpressionStatement
+
+                // selectors for the 'foo' identifier, in order of increasing specificity
+                ["*", ast.body[0].expression], // 0 identifiers, 0 pseudoclasses
+                ["ExpressionStatement > *", ast.body[0].expression], // 0 pseudoclasses, 1 identifier
+                ["Identifier", ast.body[0].expression], // 0 pseudoclasses, 1 identifier
+                [":not(Program, ExpressionStatement)", ast.body[0].expression], // 0 pseudoclasses, 2 identifiers
+                ["ExpressionStatement > Identifier", ast.body[0].expression], // 0 pseudoclasses, 2 identifiers
+                ["Identifier, ReturnStatement", ast.body[0].expression], // 0 pseudoclasses, 2 identifiers
+                ["[name = 'foo']", ast.body[0].expression], // 1 pseudoclass, 0 identifiers
+                ["[name='foo']", ast.body[0].expression], // 1 pseudoclass, 0 identifiers
+                ["ExpressionStatement > [name='foo']", ast.body[0].expression], // 1 attribute, 1 identifier
+                ["Identifier[name='foo']", ast.body[0].expression], // 1 attribute, 1 identifier
+                [":not(Program, Identifier) > [name.length=3]", ast.body[0].expression], // 1 attribute, 2 identifiers
+                ["[name='foo'][name.length=3]", ast.body[0].expression] // 2 attributes, 0 identifiers
+            ]
+        );
+    });
 
-        generator.leaveNode(dummyNode);
+    describe("parsing an invalid selector", () => {
+        it("throws a useful error", () => {
+            assert.throws(() => {
+                const emitter = new EventEmitter();
 
-        assert(emitter.emit.calledOnce);
-        assert(emitter.emit.calledWith("Foo:exit", dummyNode));
+                emitter.on("Foo >", () => {});
```

Comment:
Just to be safe, is there any way line 276 and 278 could be moved outside of `assert.throws` and only line 279 is inside the `assert.throws()` to avoid any ambiguity of which line threw?